### ISSUE TITLE
Core: "u" tag for gophercloud.BuildQueryString

### ISSUE THREAD

**Issue by ozerovandrei** [gophercloud/gophercloud#1440]
**Introduction**

Currently OpenStack Magnum won't populate `project_id` field on its standard list url:
```
https://openstack/container-infra/v1/clusters
```

When we want to retrieve the `project_id` or some additional fields we need to retrieve a detailed list by using different url:
```
https://openstack/container-infra/v1/clusters/detail
```

Python client uses `detail` option in it's `list` method:
https://github.com/openstack/python-magnumclient/blob/stable/rocky/magnumclient/v1/baseunit.py#L43
https://github.com/openstack/python-magnumclient/blob/stable/rocky/magnumclient/v1/baseunit.py#L76

It seems that it's common thing between different OpenStack services, for example Nova uses the same pattern:
```
https://openstack/compute/v2/servers/detail
https://openstack/compute/v2/servers
```

**Problems**

1. Currently Gophercloud uses an interface as an option to different List methods and we can't provide `Detailed` option as query to [gophercloud.BuildQueryString](https://github.com/gophercloud/gophercloud/blob/master/params.go#L324) because it's *not* some query option.

2. Project ID information in list response is a _must_.

**How to fix this**

1. We can introduce some common pattern to different `ToSomethingListQuery`:
```
// ToSomethingListQuery formats a ListOpts into a query string.
func (opts ListOpts) ToSomethingListQuery() (string, error) {
	q, err := gophercloud.BuildQueryString(opts)
	if opts.Detail {
		return "/detail" + q.String()
	}
	return q.String(), err
}
```

2. We can provide additional url in `service/apiversion/something/urils.go` and additional `ListDetail` method in `service/apiversion/something/requests.go`.

3. We can introduce additional tag for the [gophercloud.BuildQueryString](https://github.com/gophercloud/gophercloud/blob/master/params.go#L324) that can be used with `string` fields and will be added into [&url.URL](https://github.com/gophercloud/gophercloud/blob/master/params.go#L390) as [Path](https://golang.org/src/net/url/url.go#L344).

I personally prefer first or third variant because we won't introduce any additional functions in `requests.go` and it will look similar to different `python-something` clients that use a single `list()` function.

I also think that third variant is more elegant than first and I want to proceed with it.

@jtopjian what do you think about that?

**Comment by jtopjian** [comment]
> I personally prefer first or third variant

Ironically, the second option is the one that's already used the most 🙂 

```shell
$ grep -R Detail * | grep func | grep -v test
```

For consistency, this is the best way to go. I understand it's a little clumsy, though. 

Alternatively, I'd go with Option 1.

**Comment by ozerovandrei** [comment]
Should've grepped that 😅 

Different function totally makes sense, `detail` is not a query option after all.
And it's common sense to implement a function per controller.

Thank you for answering that. I will proceed with the second option.

**Comment by ozerovandrei** [comment]
Note: support for Magnum `detail` list will be implemented in #1442
